### PR TITLE
KAFKA-3104: add windowed aggregation to KStream

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by guozhangwang** [comment]
@ymatsuda 

1) Added the KStreamAggWindow only for tricking the generics types of KTableProcessorSupplier with Windowed<K> for getter view().

2) Make use of the WindowStore with range query, to avoid single-query per window-key. Modified its specs a bit.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamAggWindow.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            super.init(context);
+        }
+
+        @Override
+        public void process(K key, V value) {
+            // create a dummy window just for wrapping the timestamp
+            long timestamp = context().timestamp();
+
+            // send the new aggregate value
+            context().forward(new Windowed<>(key, new UnlimitedWindow(timestamp)), new Change<>(value, null));
```
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamAggWindow.java:47]: `UnlimitedWindow` ?

Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamAggWindow.java:47]: Adding this Window is just for tricking the KTableProcessorSupplier which needs to provide getter view() of Windowed<K>.

**Comment by guozhangwang** [comment]
@ymatsuda Ping for review.

Fixed unit test by adding an option to disable retaining duplicates through seqnum (this is required for aggregation).

**Comment by ymatsuda** [comment]
lgtm

**Comment by guozhangwang** [comment]
Merged to trunk. Left built-in aggregate implementations will be submitted in a different PR.

**Pull request closed without merging**